### PR TITLE
M4667 Upgrade embedded Elasticsearch 1.7.3 to standalone 5.4.1

### PR DESCRIPTION
**Pull request by dennishendriksen**
#### Checklist
- [ ] Functionality works & meets specifications
- [ ] Code reviewed
- [ ] Code unit/integration/system tested
- [ ] User documentation updated
- [ ] (If you have changed REST API interface) view-swagger.ftl updated
- [ ] Test plan template updated
- [ ] Clean commits

Related PR: https://github.com/molgenis/developer-tools/pull/3

### CODE REVIEW

**Comment by dennishendriksen** [comment]
test this please

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/index/MappingsBuilder.java [hunk 1]
```diff
@@ -88,98 +82,79 @@ private static void createAttributeMappingContents(Attribute attr, DocumentIdGen
...
+				jsonBuilder.startObject("fields").startObject(FIELD_NOT_ANALYZED).field("type", "keyword")
+						.field("index", true).endObject().endObject();
 				break;
 			case DATE_TIME:
 				jsonBuilder.field("type", "date").field("format", "date_time_no_millis");
-				// disable norms for numeric fields
-				jsonBuilder.field("norms").startObject().field("enabled", false).endObject();
 				// not-analyzed field for aggregation
-				// note: the include_in_all setting is ignored on any field that is defined in the fields options
 				// note: the norms settings defaults to false for not_analyzed fields
```
Comment (dennishendriksen) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/index/MappingsBuilder.java]: no

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/index/MappingsBuilder.java [hunk 2]
```diff
@@ -88,98 +82,79 @@ private static void createAttributeMappingContents(Attribute attr, DocumentIdGen
...
 			case TEXT:
 				// enable/disable norms based on given value
-				jsonBuilder.field("type", "string");
-				jsonBuilder.field("norms").startObject().field("enabled", enableNorms).endObject();
+				jsonBuilder.field("type", "text");
+				jsonBuilder.field("norms", true);
 				// not-analyzed field for sorting and wildcard queries
-				// note: the include_in_all setting is ignored on any field that is defined in the fields options
 				// note: the norms settings defaults to false for not_analyzed fields
 				XContentBuilder fieldsObject = jsonBuilder.startObject("fields").startObject(FIELD_NOT_ANALYZED)
```
Comment (dennishendriksen) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/index/MappingsBuilder.java]: no

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/index/MappingsBuilder.java [hunk 3]
```diff
@@ -88,98 +82,79 @@ private static void createAttributeMappingContents(Attribute attr, DocumentIdGen
...
 			case HTML:
 			case SCRIPT:
 				// enable/disable norms based on given value
-				jsonBuilder.field("type", "string");
-				jsonBuilder.field("norms").startObject().field("enabled", enableNorms).endObject();
+				jsonBuilder.field("type", "text");
+				jsonBuilder.field("norms", true);
 				// not-analyzed field for sorting and wildcard queries
-				// note: the include_in_all setting is ignored on any field that is defined in the fields options
 				// note: the norms settings defaults to false for not_analyzed fields
```
Comment (dennishendriksen) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/index/MappingsBuilder.java]: no

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/config/ElasticsearchConfig.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ */
+@Configuration
+@EnableScheduling
+@Import({ IndexConfig.class })
+public class ElasticsearchConfig
+{
+	@Value("${elasticsearch.cluster.name:molgenis}")
+	private String elasticsearchClusterName;
+
+	@Value("${elasticsearch.transport.addresses:127.0.0.1:9300}")
```
Comment (tommydeboer) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/config/ElasticsearchConfig.java]: Could you add these configuration properties to the documentation?

**Comment by fdlk** [comment]
Getting a DEBUG warning message on boot:
> [2017-06-08T14:20:40,039][INFO ][o.e.c.r.a.AllocationService] [pMOvE1t] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[baselifelines_bed0aa19][0]] ...]).
[2017-06-08T14:20:40,329][DEBUG][o.e.a.a.i.m.p.TransportPutMappingAction] [pMOvE1t] failed to put mappings on indices [[[baselifelines_bed0aa19/Ovz3JcNpS4G1I5QeeVQI5w]]], type [baselifelines_bed0aa19]
java.lang.IllegalArgumentException: Limit of nested fields [50] in index [baselifelines_bed0aa19] has been exceeded
	at org.elasticsearch.index.mapper.MapperService.checkNestedFieldsLimit(MapperService.java:586) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.index.mapper.MapperService.internalMerge(MapperService.java:443) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.index.mapper.MapperService.internalMerge(MapperService.java:334) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:266) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.applyRequest(MetaDataMappingService.java:311) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:230) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.service.ClusterService.executeTasks(ClusterService.java:633) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.service.ClusterService.calculateTaskOutputs(ClusterService.java:611) ~[elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.service.ClusterService.runTasks(ClusterService.java:570) [elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.service.ClusterService$ClusterServiceTaskBatcher.run(ClusterService.java:262) [elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:569) [elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:247) [elasticsearch-5.4.1.jar:5.4.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:210) [elasticsearch-5.4.1.jar:5.4.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_91]

**Comment by fdlk** [comment]
Mapping service performance somewhat degraded, I think:
### before
screen shot 2017-06-08 at 14 40 51
### after
screen shot 2017-06-08 at 14 42 39

File: molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/config/ElasticsearchConfig.java (hunk 4 above)
Comment (dennishendriksen) [inline molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/config/ElasticsearchConfig.java]: Done

**Comment by dennishendriksen** [comment]
@fdlk index.mapping.nested_fields.limit increased from 50 to Long.MAX_VALUE
@fdlk I propose we accept the slight degradation in mapping results

**Comment by fdlk** [comment]
Should move the recommended elasticsearch security settings described in the developer security docs from tomcat options to elasticsearch config.

**Comment by dennishendriksen** [comment]
@fdlk I don't understand what you mean. Lets have a chat :-)

**Pull request merged**